plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,514 @@
+	chunks_isp =
+	    fu_chunk_array_new(isp_data, isp_data_size, address, 0x0, 64 - basic_data_size);
+
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks_isp->len);
+
+	for (guint i = 0; i < chunks_isp->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks_isp, i);
+		g_autoptr(GByteArray) part_isp_data = fu_struct_algoltek_cmd_transfer_pkt_new();
+
+		fu_struct_algoltek_cmd_transfer_pkt_set_len(part_isp_data,
+							    basic_data_size +
+								fu_chunk_get_data_sz(chk));
+		fu_struct_algoltek_cmd_transfer_pkt_set_cmd(part_isp_data, FU_ALGOLTEK_CMD_ISP);
+		fu_struct_algoltek_cmd_transfer_pkt_set_address(part_isp_data,
+								fu_chunk_get_address(chk));
+		if (!fu_struct_algoltek_cmd_transfer_pkt_set_data(part_isp_data,
+								  fu_chunk_get_data(chk),
+								  fu_chunk_get_data_sz(chk),
+								  error)) {
+			g_prefix_error(error, "assign isp data failure: ");
+			return FALSE;
+		}
+		fu_struct_algoltek_cmd_transfer_pkt_set_checksum(
+		    part_isp_data,
+		    ~fu_sum8(part_isp_data->data, part_isp_data->len) + 1);
+
+		if (!algoltek_device_ctrl_transfer(self,
+						   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+						   FU_ALGOLTEK_CMD_ISP,
+						   0,
+						   0,
+						   part_isp_data,
+						   part_isp_data->data[0],
+						   error)) {
+			g_prefix_error(error, "isp failure: ");
+			return FALSE;
+		}
+		fu_progress_step_done(progress);
+	}
+
+	return TRUE;
+}
+
+static gboolean
+algoltek_device_bot(FuAlgoltekUsbDevice *self, int address, GError **error)
+{
+	g_autoptr(GByteArray) buf = fu_struct_algoltek_cmd_address_pkt_new();
+
+	fu_struct_algoltek_cmd_address_pkt_set_len(buf, 5);
+	fu_struct_algoltek_cmd_address_pkt_set_cmd(buf, FU_ALGOLTEK_CMD_BOT);
+	fu_struct_algoltek_cmd_address_pkt_set_address(buf, address);
+	fu_struct_algoltek_cmd_address_pkt_set_checksum(buf, ~fu_sum8(buf->data, buf->len) + 1);
+
+	if (!algoltek_device_ctrl_transfer(self,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   FU_ALGOLTEK_CMD_BOT,
+					   0,
+					   0,
+					   buf,
+					   buf->data[0],
+					   error)) {
+		g_prefix_error(error, "system boot failure: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+algoltek_device_ers(FuAlgoltekUsbDevice *self, GError **error)
+{
+	g_autoptr(GByteArray) buf = fu_struct_algoltek_cmd_address_pkt_new();
+
+	fu_struct_algoltek_cmd_address_pkt_set_len(buf, 3);
+	fu_struct_algoltek_cmd_address_pkt_set_cmd(buf, FU_ALGOLTEK_CMD_ERS);
+	fu_struct_algoltek_cmd_address_pkt_set_checksum(buf, ~fu_sum8(buf->data, buf->len) + 1);
+
+	if (!algoltek_device_ctrl_transfer(self,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   FU_ALGOLTEK_CMD_ERS,
+					   0,
+					   0,
+					   buf,
+					   buf->len,
+					   error)) {
+		g_prefix_error(error, "data clear failure: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_device_status_check(FuDevice *self, gpointer user_data, GError **error)
+{
+	guint16 update_status;
```

Comment:
I'm confused how this can be a uint16

Reply:
Ohh, but I see:

    update_status = update_status_array->data[0];

i.e. it can only ever be one byte. Should that have been `fu_memread_uint16_safe` on the buffer instead?

Reply:
We can try this one and see how we can adopt it.

Reply:
Updated! Revise the type from guint16 to guint8.